### ISSUE TITLE
How to train a 2-stem model, separating vocals and accompaniment

### ISSUE THREAD

**Issue by zhouyong64** [facebookresearch/demucs#176]
The paper and the code are for training 4-stem models. Now I just want to separate vocals from accompaniment, 2-stem models. What changes do I need to do, for example in terms of configuration and training parameters? I couldn't find any document on this. Thanks.

**Comment by adefossez** [comment]
This is not directly supported without code modification at the moment. You would have to wait on an update, or adapt the code yourself (you will need sufficient knowledge in python to debug this yourself):
- https://github.com/facebookresearch/demucs/blob/master/demucs/__main__.py#L119 replace `sources=SOURCES` by `sources=['accompaniment', 'vocals']`.
- https://github.com/facebookresearch/demucs/blob/master/demucs/train.py#L57: add `import torch` at the top of the file, and add after line 57: `sources = torch.stack([sources[:, :3].sum(1), sources[:, 3]], dim=1)`
- https://github.com/facebookresearch/demucs/blob/master/demucs/train.py#L118 add the same thing, just after, `sources = torch.stack([sources[:, :3].sum(1), sources[:, 3]], dim=1)`

i hope this helps.

**Comment by lyghter** [comment]
Hello @adefossez
I want to separate only one stem, e.g. vocals. Did I change the code correctly? Will a 1-stem model perform faster on training/inference then 2-stem and 4-stem models?
* https://github.com/facebookresearch/demucs/blob/master/demucs/__main__.py#L119 replace `sources=SOURCES` by `sources=['vocals']`
* https://github.com/facebookresearch/demucs/blob/master/demucs/train.py#L57: add `import torch` at the top of the file, and add after line 57: `sources = torch.stack([sources[:, 3]], dim=1)`
* https://github.com/facebookresearch/demucs/blob/master/demucs/train.py#L118 add the same thing, just after, `sources = torch.stack([sources[:, 3]], dim=1)`

**Comment by Sushmitha558** [comment]
are these only changes needed to train for 2 stem models? Also not able to find sources=SOURCES in https://github.com/facebookresearch/demucs/blob/master/demucs/__main__.py#L119

**Comment by Sushmitha558** [comment]
> are these only changes needed to train for 2 stem models? Also not able to find sources=SOURCES in https://github.com/facebookresearch/demucs/blob/master/demucs/__main__.py#L119
